src/v/datalake/tests/datalake_protobuf_tests.cc

```diff
@@ -0,0 +1,559 @@
+        BrokerShard broker_shard;
+        broker_shard.set_id(bs.first);
+        broker_shard.set_shard(bs.second);
+        ret.mutable_replicas()->Add(std::move(broker_shard));
+    }
+
+    return ret;
+}
+
+TEST_CORO(values_protobuf, TestComplexValueConversion) {
+    Metadata message;
+    message.mutable_nodes()->Add(11);
+    message.mutable_nodes()->Add(12);
+    message.mutable_nodes()->Add(13);
+    Topic topic_1;
+    topic_1.set_name("topic_1");
+
+    topic_1.mutable_partitions()->Add(
+      make_partition(0, {{11, 1}, {12, 0}, {13, 4}}));
+    topic_1.mutable_partitions()->Add(
+      make_partition(1, {{12, 1}, {11, 0}, {13, 4}}));
+    topic_1.mutable_partitions()->Add(
+      make_partition(3, {{11, 3}, {12, 1}, {13, 2}}));
+
+    message.mutable_topics()->emplace("topic_1", std::move(topic_1));
+
+    Topic topic_2;
+    topic_2.set_name("topic_2");
+    topic_2.mutable_partitions()->Add(make_partition(0, {{11, 0}}));
+    message.mutable_topics()->emplace("topic_2", std::move(topic_2));
+
+    auto result = co_await serialize_and_convert(message);
+
+    ASSERT_TRUE_CORO(result.has_value());
+    auto optional_value = std::move(result.value());
+    ASSERT_TRUE_CORO(result.value().has_value());
+    auto result_value = std::get<std::unique_ptr<struct_value>>(
+      std::move(optional_value.value()));
+
+    EXPECT_THAT(
+      result_value->fields,
+      ElementsAre(
+        IcebergMap(UnorderedElementsAre(
+          IcebergKeyValue(
+            IcebergPrimitive<string_value>("topic_2"),
+            // topic
+            IcebergStruct(
+              OptionalIcebergPrimitive<string_value>("topic_2"),
+              IcebergList(ElementsAre(
+                // partition
+                IcebergStruct(
+                  OptionalIcebergPrimitive<int_value>(0),
+                  IcebergList(ElementsAre(
+                    // broker shard
+                    IcebergStruct(
+                      OptionalIcebergPrimitive<iceberg::long_value>(11),
+                      OptionalIcebergPrimitive<iceberg::int_value>(0)))),
+                  OptionalIcebergPrimitive<int_value>(0)))))),
+          IcebergKeyValue(
+            IcebergPrimitive<string_value>("topic_1"),
+            // topic
+            IcebergStruct(
+              OptionalIcebergPrimitive<string_value>("topic_1"),
+              IcebergList(ElementsAre(
+                // partition
+                IcebergStruct(OptionalIcebergPrimitive<int_value>(0), _, _),
+                // partition
+                IcebergStruct(OptionalIcebergPrimitive<int_value>(1), _, _),
+                // partition
+                IcebergStruct(
+                  OptionalIcebergPrimitive<int_value>(3), _, _))))))),
+        IcebergList(ElementsAre(
+          OptionalIcebergPrimitive<long_value>(11),
+          OptionalIcebergPrimitive<long_value>(12),
+          OptionalIcebergPrimitive<long_value>(13)))));
+}
+
+TEST_CORO(values_protobuf, TestEmptyMessage) {
```

Comment:
Thank you for these extra tests - just one more test and I think we're good to go.

Can we test that a set, but otherwise empty message is represented correctly? Example:

```proto
message Foo {
  message Nested {
    string bar = 1;
  }
  Nested nested = 1;
}
```

```cpp
Foo foo;
foo.mutable_nested(); // set foo, but don't set any fields in foo

// This should parse into a struct with a single field that is an empty struct.
```

Reply:
I've added a test based on the proto definitions in `TestAllTypesEdition2023`

Reply:
Updated proto:

```proto
message Foo {
  message Nested {
    message NestedAgain {
      string bar = 1;
    }
    NestedAgain qux = 1;
  }
  Nested nested = 1;
}
```

Reply:
done